core/src/test/java/org/elasticsearch/action/support/replication/TransportWriteActionTests.java

```diff
@@ -222,6 +402,80 @@ protected TestResponse newResponseInstance() {
         }
     }
 
+    final IndexService mockIndexService(final IndexMetaData indexMetaData, ClusterService clusterService) {
+        final IndexService indexService = mock(IndexService.class);
+        when(indexService.getShard(anyInt())).then(invocation -> {
+            int shard = (Integer) invocation.getArguments()[0];
+            final ShardId shardId = new ShardId(indexMetaData.getIndex(), shard);
+            if (shard > indexMetaData.getNumberOfShards()) {
+                throw new ShardNotFoundException(shardId);
+            }
+            return mockIndexShard(shardId, clusterService);
+        });
+        return indexService;
+    }
+
+    final IndicesService mockIndicesService(ClusterService clusterService) {
```

Comment:
can we share this and mockIndexShard with TransportReplicationAction? is there any point in that?